#' Read in exon annotation
#'
#' Reads in an exon annotation file generated by gtfToExons
#' Alternatively, will generate then read the exon annotation file if a .gtf file and an ouput file is specified
#' @param exonFile exon annotation file generated by gtfToExons
#' @param outputFile name of the file to write the exon annotation to when a .gtf file is specified.
#' @return exon annotation data.frame
#' @export
#' @examples
#' gtf <- system.file("extdata","gencode.v24.annotation.small.gtf",
#' package = "branchpointer")
#' exons <- readExonAnnotation(gtf)
#'
#' small_exons <- system.file("extdata","gencode.v24.annotation.exons.small.txt",
#' package = "branchpointer")
#' exons <- readExonAnnotation(small_exons)
#' @author Beth Signal

readExonAnnotation <- function(exonFile, outputFile=NA){

  if(stringr::str_sub(exonFile,-4,-1) == ".gtf"){
    exons <- gtfToExons(exonFile, outputFile=outputFile)
  }else{
    exons <- as.data.frame(data.table::fread(exonFile, stringsAsFactors = FALSE))
  }
  colnames(exons) <- c("chromosome","type","start","end","strand",
                       "gene_id","gene_type","transcript_id",
                       "transcript_type","exon_number","exon_id")
  return(exons)
}
