#' Read in exon annotation
#'
#' Reads in an exon annotation file generated by gtfToExons
#' Alternatively, will generate then read the exon annotation file if a .gtf file is specified
#' @param exon_file exon annotation file generated by gtfToExons
#' @param output_file name of the file to write the exon annotation to when a .gtf file is specified.
#' If not specified, will write to the same location as the gtf file.
#' @return exon annotation data.frame
#' @export
#' @examples
#' exons <- readExonAnnotation("gencode.v24.annotation.exons.txt")
#' exons <- readExonAnnotation("gencode.v24.annotation.gtf")
#' @author Beth Signal

readExonAnnotation <- function(exon_file, output_file=NA){

  if(stringr::str_sub(exon_file,-4,-1) == ".gtf"){
    if(is.na(output_file)){
      output_file=gsub(".gtf", ".exons.txt", exon_file)
    }

    exons <- gtfToExons(exon_file, output_file=output_file)

  }else{
    exons <- as.data.frame(data.table::fread(exon_file, stringsAsFactors = FALSE))
  }
  colnames(exons) <- c("chromosome","type","start","end","strand",
                       "gene_id","gene_type","transcript_id",
                       "transcript_type","exon_number","exon_id")
  return(exons)
}
